fairscale/nn/model_parallel/initialize.py

```diff
@@ -88,32 +89,119 @@ def initialize_model_parallel(
+        for j in range(context_parallel_size):
+            for k in range(model_parallel_size):
+                ranks = groups[i, :, j, k].tolist()
+                group = torch.distributed.new_group(ranks, backend=pipeline_backend)
+                if i == found[0] and j == found[2] and k == found[3]:
+                    _PIPELINE_PARALLEL_GROUP = group
+                    _PIPELINE_PARALLEL_RANKS = ranks
+
+
+    # Build the context parallel groups.
+    global _CONTEXT_PARALLEL_GROUP
+    global _CONTEXT_PARALLEL_GROUP_RANKS
+    global _CP_PREV_RANK
+    global _CP_NEXT_RANK
+    global _CP_ZERO_RANK
+
+    assert (
+        _CONTEXT_PARALLEL_GROUP is None
+    ), "Context parallelism is already initialized."
+    for i in range(data_parallel_size):
+        for j in range(pipeline_length):
+            for k in range(model_parallel_size):
+                ranks = groups[i, j, :, k].tolist()
+                group = torch.distributed.new_group(ranks, backend=cp_backend)
+                if i == found[0] and j == found[1] and k == found[3]:
+                    _CONTEXT_PARALLEL_GROUP = group
+                    _CONTEXT_PARALLEL_GROUP_RANKS = ranks
+                    index = ranks.index(rank)
+                    _CP_PREV_RANK = ranks[(index + 1) % len(ranks)]
+                    _CP_NEXT_RANK = ranks[index - 1]
+                    _CP_ZERO_RANK = ranks[0]
+
+
+def model_parallel_is_initialized() -> bool:
+    """Check if model and data parallel groups are initialized."""
+    if _MODEL_PARALLEL_GROUP is None or _DATA_PARALLEL_GROUP is None or _PIPELINE_PARALLEL_GROUP is None:
+        return False
+    return True
+
+
+def get_context_parallel_group():
+    """Get the context parallel group the caller rank belongs to."""
+    assert (
+        _CONTEXT_PARALLEL_GROUP is not None
+    ), "context parallel group is not initialized"
+    return _CONTEXT_PARALLEL_GROUP
+
+def get_context_parallel_rank():
+    assert (
+        _CONTEXT_PARALLEL_GROUP is not None
+    ), "context parallel group is not initialized"
+    return torch.distributed.get_rank(group=_CONTEXT_PARALLEL_GROUP)
+
+
+def get_context_parallel_prev_rank():
+    assert (
+        _CONTEXT_PARALLEL_GROUP is not None
+    ), "context parallel group is not initialized"
+    return _CP_PREV_RANK
+
+
+def get_context_parallel_next_rank():
+    assert (
+        _CONTEXT_PARALLEL_GROUP is not None
+    ), "context parallel group is not initialized"
+    return _CP_NEXT_RANK
+
+
+def get_context_parallel_src_rank():
+    assert (
+        _CONTEXT_PARALLEL_GROUP is not None
+    ), "context parallel group is not initialized"
+    return _CP_ZERO_RANK
+
+
+def get_context_parallel_world_size():
+    """Return world size for the context parallel group."""
+    if torch.distributed.is_available() and torch.distributed.is_initialized():
+        return torch.distributed.get_world_size(group=get_context_parallel_group())
+    else:
+        return 0
```

Comment:
is 0 right default value? usually world size returns 1 by default?

Reply:
We should return torch.distributed.get_world_size(...) directly here. `get_context_parallel_group()` already checks the CP pg is initialized.